libz-rs-sys/src/gz.rs

```diff
@@ -1296,6 +1303,340 @@ unsafe fn gz_decomp(state: &mut GzState) -> Result<(), ()> {
+    }
+
+    // Allocate memory if this is the first time through.
+    if state.input.is_null() && gz_init(state).is_err() {
+        return 0;
+    }
+
+    /* FIXME uncomment once seek support is implemented
+    if state.seek {
+        state.seek = 0;
+        if gz_zero(state, state.skip) == -1 {
+            return 0;
+        }
+    }
+     */
+
+    let put = len as c_int;
+
+    // For small len, copy to input buffer, otherwise compress directly.
+    if (len as usize) < state.in_size {
+        // Copy to input buffer, compress when full.
+        loop {
+            if state.stream.avail_in == 0 {
+                state.stream.next_in = state.input;
+            }
+            let have = unsafe {
+                state
+                    .stream
+                    .next_in
+                    .add(state.stream.avail_in as usize)
+                    .offset_from(state.input)
+            } as usize;
+            let copy = cmp::min(state.in_size.saturating_sub(have), len as usize);
+            // Safety: The caller is responsible for ensuring that buf points to at least len readable
+            // bytes, and copy is <= len.
+            unsafe { ptr::copy(buf, state.input.add(have).cast::<c_void>(), copy) };
+            state.stream.avail_in += copy as c_uint;
+            state.pos += copy as u64;
+            buf = unsafe { buf.add(copy) };
+            len -= copy as c_uint;
+            if len != 0 && gz_comp(state, Z_NO_FLUSH).is_err() {
+                return 0;
+            }
+            if len == 0 {
+                break;
+            }
+        }
+    } else {
+        // Consume any data left in the input buffer.
+        if state.stream.avail_in != 0 && gz_comp(state, Z_NO_FLUSH).is_err() {
+            return 0;
+        }
+
+        // Directly compress user buffer to file.
+        state.stream.next_in = buf.cast::<_>();
+        loop {
+            let n = cmp::min(len, c_uint::MAX);
+            state.stream.avail_in = n;
+            state.pos += n as u64;
+            if gz_comp(state, Z_NO_FLUSH).is_err() {
+                return 0;
+            }
+            len -= n;
+            if len == 0 {
+                break;
+            }
+        }
+    }
+
+    // Input was all buffered or compressed.
+    put
+}
+
+// Initialize `state` for writing a gzip file.  Mark initialization by setting
+// `state.input` to non-null.
+//
+// # Returns
+//
+// - `Ok` on success.
+// - `Err` on error.
+fn gz_init(state: &mut GzState) -> Result<(), ()> {
+    // Allocate input buffer (double size for gzprintf).
+    state.in_size = state.want * 2;
+    let Some(input) = ALLOCATOR.allocate_slice_raw::<u8>(state.in_size) else {
+        // Safety: The caller confirmed the validity of state.
+        unsafe { gz_error(state, Some((Z_MEM_ERROR, "out of memory"))) };
+        return Err(());
+    };
+    state.input = input.as_ptr();
+    // Note: zlib-ng fills the input buffer with zeroes here, but it's unneeded.
```

Comment:
Are you sure there won't be any issues with reading uninitialized memory?

Reply:
The `gz` code only reads parts of `input` that have been written (confirmed by Miri). As long as the `inflate` and `deflate` code don't read past the filled part of `input`, we won't need to zero out the memory.

Reply:
They won't read past the range that you give it. assuming you don't give it uninitialized input bytes we should be good then.